app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php

```diff
@@ -0,0 +1,506 @@
+                    } else {
+                        continue;
+                    }
+                }
+            }
+            if ($leadCompany instanceof Lead && 'lead' === $field->getObject()) {
+                $leadField = $leadCompany->getField($field->getAlias());
+                if (false === $leadField) {
+                    continue;
+                }
+
+                $field     = $this->fieldModel->getRepository()->find($leadField['id']);
+
+                if (null === $field) {
+                    continue;
+                }
+            }
+            if (false === $leadField) {
+                continue;
+            }
+
+            $leadsCompanies[$key] = $this->setHash($leadCompany, $leadField, $field, $pseudonymize);
+        }
+
+        return $leadsCompanies;
+    }
+
+    /**
+     * @param array<string, string|null> $field
+     */
+    private function setHash(
+        Company|Lead $leadOrCompany,
+        array $field,
+        LeadField $leadField,
+        bool $pseudonymize,
+    ): Lead|Company {
+        if (empty($field['value'])) {
+            return $leadOrCompany;
+        }
+
+        try {
+            if ('email' === $field['type']) {
+                $valueAnonymized = AnonymizeHelper::email($field['value'], $pseudonymize);
+                if ($leadField->getCharLengthLimit() < strlen($valueAnonymized)) {
+                    $valueAnonymized = $this->formatHashEmail($valueAnonymized, $leadField->getCharLengthLimit());
+                }
+                $this->updateEmailStatusValues($field['value'], $valueAnonymized, $pseudonymize);
+            } else {
+                $valueAnonymized = AnonymizeHelper::text($field['value'], $pseudonymize);
+            }
+
+            if ($leadField->getCharLengthLimit() < strlen($valueAnonymized) && 'email' === $field['type']) {
+                $valueAnonymized = $this->formatHashEmail($valueAnonymized, $leadField->getCharLengthLimit());
+            } elseif ($leadField->getCharLengthLimit() < strlen($valueAnonymized)) {
+                $valueAnonymized = substr($valueAnonymized, 0, $leadField->getCharLengthLimit());
+            }
+
+            if ($leadOrCompany instanceof Lead) {
+                $leadOrCompany->addUpdatedField($leadField->getAlias(), $valueAnonymized);
+            }
+
+            if ($leadOrCompany instanceof Company) {
+                $alias       = self::COMPANY_FIELDS_TO_COLUMNS[$leadField->getAlias()] ?? $leadField->getAlias();
+                if (property_exists($leadOrCompany, $alias)) {
+                    $leadOrCompany->{'set'.ucfirst($alias)}($valueAnonymized);
+                }
+            }
+        } catch (\Exception $e) {
+            // Do nothing
+            $this->logger->error('AnonymizeUserDataSubscriber setHash fail: '.$e->getMessage());
+        }
+
+        return $leadOrCompany;
+    }
+
+    private function formatHashEmail(string $email, int $limit): string
```

Comment:
Consider delegating this logic to a service class to improve the separation of concerns.